### ISSUE TITLE
Support OS appearance events on GNOME

### ISSUE THREAD

**Issue by EpocSquadron** [wez/wezterm#1098]
I'd really like to be able to switch terminal theme with the rest of my UI in the context of GNOME desktop on linux. This currently does not work in either X11 or Wayland.

I'm not sure how this can be acheived, but I know that Firefox is able to detect the appearance change and even propagates it to webpages with the prefers-color-scheme CSS media query. I do know that Firefox makes use of the GTK toolkit to provide GNOME integration more effectively though. I know that's not the case here (which is also why client side decorations are out of place when run in wayland mode in GNOME).

Despite that knowledge, I tracked down some discussion on the implementation of dark mode detection in webkit, which contains C code that might be useful. https://bugs.webkit.org/show_bug.cgi?id=196685

Besides that, we can query gsettings if present at window creation and match the theme that way. https://wiki.archlinux.org/title/Dark_mode_switching#gsettings

Either way, dark mode in GTK appears to be implemented as paired themes. Themes with a (-dark) variant would work with either of the above methods as the heuristic is really "is -dark in the name of the current theme". Other themes or themes that are only dark or light will not work with the detection.

For the window event, I don't know if there's any method to listen for theme changes in GTK without being a GTK app, but I also didn't research that.

**Comment by lunaryorn** [comment]
I came across this, and would like to point out that the upcoming Gnome 42 release provides the dark mode preference through the generic XDG settings portal, which in addition to Gnome is going to be supported by at least KDE and elementary as well.

This portal should work consistently across X11 and Wayland, and provides the current setting as well as a change signal over DBus, see <https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.Settings>

I think this would allow implementing `window:get_appearance()` on Wayland with no dependencies other than DBus, and given that wezterm does notifications on Linux I presume it already uses DBus one way or another.  The change signal would be a bit of a challenge I guess, because receiving DBus signals usually needs an open connection to the bus and some kind of mainloop.

@wez  Would you accept a pull request which implements `window:get_appearance` for Wayland based on the XDG portal API, and perhaps the `window

**Comment by wez** [comment]
Yeah, I'd like to see that!  https://github.com/wez/wezterm/blob/main/wezterm-toast-notification/src/dbus.rs is our integration with dbus: it does listen for events to be able to handle things like clicking to open up the changelog URL when we have a release update.  It uses the `zbus` crate for that.

**Comment by lunaryorn** [comment]
@wez Ah cool, I know zbus already.  I'll look at it :pray:

**Comment by lunaryorn** [comment]
@wez I've looked at making wezterm use the XDG settings portal to detect Gnome/KDE dark mode, and wondered whether you would answer a few questions about WezTerm's code?

I saw that the macOS and Windows frontends get a native event on the window connection and then use `self.event.                          .dispatch(WindowEvent::AppearanceChanged(…))` to forward this event to WezTerm's internal event handling.  I looked at the `.dispatch` machinery, and it doesn't seem to do any cross-thread magic, so I presume on Windows and macOS the `.dispatch` call and thus also the `config_was_reloaded` call and ultimately the Lua event handlers for the `window-config-reloaded` event run on the GUI main thread.

However, the portal sends signals over DBus, i.e. a separate connection, which presumably I'd need to handle in a separate thread, so I guess I'd somehow need to dispatch an event on WezTerm's main thread in response a DBus signal?  How can I do that in WezTerm?  Or can I directly dispatch a futures `Stream` on WezTerm's main loop?
